Slot-2 addon support (v1.0)

BueniaDev opened the pull request:
This pull request adds support for the DS Rumble Pak (packaged with Metroid Prime Pinball, compatible with several DS games), the Guitar Grip (required for the Guitar Hero: On Tour games), and the DS Memory Expansion Pak (required for the DS Browser).

Located in the emulator settings dialog is a drop-down menu (below the "Console type" drop-down) for choosing the desired Slot-2 addon.
The "Rumble Pak" option emulates the DS Rumble Pak, the "Guitar Grip" option emulates the Guitar Grip used by the Guitar Hero: On Tour games, the "Memory Expansion Pak" option emulates the DS Memory Expansion Pak, and the "None" option emulates no Slot-2 addon.

The Guitar Grip implementation comes with assignable hotkeys for the Guitar Grip's four colored buttons, which can be configured in the "Addons" tab of the input config dialog.

I intend on solving most, if not all, of the add-on related issues in mainstream melonDS, which is why this proposed Slot-2 add-on implementation constitutes the initial version of many. As such, if this pull request is merged, expect support for more add-ons, such as the "Tony Hawk" motion pak, the "Slide Controller", and others, to be added in future versions.

I'm opening this pull request early for any feedback related to coding style and UI design.

Here is a demo of the Memory Expansion Pak implementation:

https://drive.google.com/file/d/1l1PTeNzFacDgqN5hr5-PmjXi2T_rE_iu/view?usp=sharing

Demos for the other emulated add-ons are available upon request.

rzumer commented on src/NDS.cpp:
```diff
@@ -2073,11 +2073,6 @@ u32 ARM9Read32(u32 addr)
         
-        if (addr == 0x80000AC)
-        {
-        		printf("ARM9 Read16: 0x%08x\n", ARM9->R[15]);
```
Why is this removed?

Buenia0 replied:
That was unnecessary logic used to debug my implementation.

rzumer replied:
Okay, then I guess it's best to remove it from the commit that introduced it via interactive rebase (like other minor formatting issues).

rzumer commented on src/GBACart.h:
```diff
@@ -1,87 +1,87 @@
-
-#endif // GBACART_H
+/*
+    Copyright 2019 Arisotura, Buenia0
```
I think you modified the wrong header, since I wrote this one, just not the Slot2 one.

Buenia0 replied:
Honesst mistake, rzumer! I'll change it back!

rzumer commented on src/frontend/qt_sdl/EmuSettingsDialog.cpp:
```diff
@@ -151,6 +162,7 @@ void EmuSettingsDialog::done(int r)
         std::string dsiSDPath = ui->txtDSiSDPath->text().toStdString();
 
         if (consoleType != Config::ConsoleType
+        		|| chosenaddon != Config::Slot2Addon
```
This seems misaligned (has an extra tab?)

rzumer commented on src/frontend/qt_sdl/Input.cpp:
```diff
@@ -65,17 +68,51 @@ void OpenJoystick()
+{
+    if (Haptic != nullptr)
+    {
+    		isFrameDone = false;
```
Stray tab here.

rzumer commented on src/frontend/qt_sdl/Input.cpp:
```diff
@@ -65,17 +68,51 @@ void OpenJoystick()
+{
+    if (Haptic != nullptr)
+    {
+    		isFrameDone = true;
```
Please convert tabs to spaces (4 spaces/tab).

rzumer commented on src/frontend/qt_sdl/main.cpp:
```diff
@@ -507,6 +516,14 @@ void EmuThread::run()
+            // implementation to take into account variable framerates...
+            if (Input::isFrameDone)
+            {
+            		Platform::StopRumble();
```
Tabs here, should be converted to spaces.

RSDuck commented on src/Slot2Cart.h:
```diff
@@ -0,0 +1,78 @@
+}
+
+
+enum GuitarKeys : int
```
please use a prefixed old fashioned enum here instead

rzumer commented on src/frontend/qt_sdl/main.cpp:
```diff
@@ -507,6 +516,14 @@ void EmuThread::run()
+            
+            // This piece of code is used in order for the Rumble Pak
+            // implementation to take into account variable framerates...
+            if (Input::isFrameDone)
```
I don't know if we even need `IsFrameDone`, since it's set to true whenever rumble is set. You can probably just call `StopRumble()` after every frame based on your current implementation. Anyway the variable name doesn't make sense to me, since it seems to indicate that rumble should be stopped once the frame is done (rather than the frame actually being done).

Besides that, if you've tested it and it feels right to you I'm fine with that implementation. I don't want to delay this until I get around to comparing it to real hardware, but I'd like a TODO added here to remind developers that testing is needed.

rzumer commented on src/frontend/qt_sdl/PlatformConfig.cpp:
```diff
@@ -112,6 +113,11 @@ ConfigEntry PlatformConfigFile[] =
+    {"HKKey_GuitarGripRed",       0, &HKKeyMapping[HK_GuitarGripRed],       -1, NULL, 0},
+    {"HKKey_GuitarGripYellow",    0, &HKKeyMapping[HK_GuitarGripYellow],    -1, NULL, 0},
+    {"HKKey_GuitarGripBlue",      0, &HKKeyMapping[HK_GuitarGripBlue],      -1, NULL, 0},
+
```
Stray newline here.

Buenia0 replied:
Currently working on resolving that issue right now. Should be ready later today!

Buenia0 replied:
I've already resolved this issue, rzumer!

rzumer replied:
I still see one unnecessary line break added between this block and the next.

rzumer commented on src/frontend/qt_sdl/Platform.cpp:
```diff
@@ -411,5 +414,14 @@ int LAN_RecvPacket(u8* data)
 
+void StartRumble()
+{
+    Input::StartRumble();
```
Since those are one-line functions now, `Input::StartRumble()` and `Input::StopRumble()` can be called directly and these intermediaries removed.

rzumer commented on src/frontend/qt_sdl/EmuSettingsDialog.cpp:
```diff
@@ -196,6 +208,9 @@ void EmuSettingsDialog::done(int r)
             Config::DirectBoot = directBoot;
+            
+            Config::Slot2Addon = chosenaddon;
+    				NDS::SetSlot2Addon(Config::Slot2Addon);
```
Tabs here to convert.

Buenia0 replied:
Almost forgot about that! Thanks for reminding me!

rzumer commented on src/Slot2Cart.cpp:
```diff
@@ -0,0 +1,213 @@
+    // This should allow commercial games to properly detect the Rumble Pak.
+    // Credit to rzumer for coming up with this algorithm...
+    
+    u8 lodata = ((addr | 0x40) & 0xFFFD);
```
This should be u16. The upper byte doesn't seem to matter anyway, but it matches GBATEK more and if we're returning a 16-bit value that's mostly open bus there's no reason to clear the upper byte.

rzumer commented on src/frontend/qt_sdl/EmuSettingsDialog.cpp:
```diff
@@ -133,6 +143,7 @@ void EmuSettingsDialog::done(int r)
 
         int consoleType = ui->cbxConsoleType->currentIndex();
         int directBoot = ui->chkDirectBoot->isChecked() ? 1:0;
+        int chosenaddon = ui->cbxSlot2Addon->currentIndex();
```
This needs to be camel-cased.

rzumer commented on src/frontend/qt_sdl/Input.cpp:
```diff
@@ -36,6 +37,8 @@ u32 HotkeyPress, HotkeyRelease;
 
 u32 InputMask;
 
+bool isFrameDone = false;
```
This should be Pascal-cased (though as mentioned in another comment I don't know if it's needed at all).

literalmente-game commented:
While this makes the Guitar Hero games reach in-game, they're still not very playable (at least not in a _confortable_ way). You need to strum on the guitar (center of the touchscreen) for each note, so you need to hold the note button with a key (or a button on the controller) while making the touch gesture with the mouse... really weird to play like this (and certainly 100% impossible to play any song on hard as you need to strum really fast).

One possible solution to cover at least all the Guitar Hero games: shortcut to key/buttons for the touchscreen motion needed, so you can play it properly with both controllers and keyboards

RSDuck commented:
I doubt that adding hardcoded touch macros is in long term a good solution.

literalmente-game commented:
Agreed, this was the first thing that came to mind as a "dirt quick" fix. As an actual solution (or at least close to it), Citra latest report came to mind with their mentioned "Touchscreen Mappings". Details here https://github.com/citra-emu/citra/pull/5163

RSDuck commented:
thanks for the reference, that's certainly interesting. The Switch frontend already has some analog stick -> touchscreen mapping options, which I want bring over to the main repository.

SA-X0 commented:
Question.
Is Slot-2 addon support coming back or not anymore?
Before the SD support got added the Slot-2 addon support disappeared then was back again in the next builds.
After this build https://github.com/Arisotura/melonDS/actions/runs/230375997 the Slot-2 addon support hasn't been added anymore.
https://imgbox.com/g/1gZOL02iOU
Thanks melonDS devs.

RSDuck commented:
it's because that commit made change which are in conflict with those of this PR.

SA-X0 commented:
Too bad.
I'll keep a separate build for this great feature in the meantime, i like the rumble pak.
Thanks RSDuck.

Buenia0 commented:
Alright, just fixed the merge conflicts. Should be good now!

SA-X0 commented:
Yay!
Thank you Buenia0.

Buenia0 commented:
You're welcome, SA-X0!

SA-X0 commented:
Seems the other branches don't like this one, recent build again Slot-2 addon support is gone.

rzumer commented:
There are still some unresolved comments, make sure to expand collapsed discussions. You should be able to mark ones that were dealt with as resolved to collapse them by default.

Buenia0 commented on src/GBACart.cpp:
```diff
@@ -1,773 +1,963 @@
+    {
+        Platform::StopRumble();
+        RumbleState = val;
+        Platform::StartRumble(16);
```
Hello, rzumer. Have you been able to access your hardware and test the different Rumble Pak implemtations yet? If so, what were your findings? I just need to know that so I can determine how to properly implement the Rumble Pak's pulse timing!

BueniaDev replied:
Okay, if no one's gonna reply to this issue, then I'm just gonna mark it as resolved...

Buenia0 commented:
Alright, just resolved most of the unresolved comments. I'm still waiting on the rest until we have more information on those fronts. Hope this gets merged soon!

Buenia0 commented:
I think I know why the Ubuntu aarch64 build is failing. It's because of an installation issue with the grub-efi-amd64-signed package. I have no idea what's causing that error, so if anyone has any ideas as to why that's the case, please let me know!

rzumer commented on src/Slot2Cart.h:
```diff
@@ -0,0 +1,78 @@
+void DoSavestate(Savestate* file);
+}
+
+#endif // SLOT2CART_H
```
There is still no newline at the end.

Buenia0 replied:
I thought I added a trailing newline to the end in one of my most recent commits. Here is proof that I did that:

![image](https://user-images.githubusercontent.com/32279537/93609352-5b909e80-f991-11ea-98d8-219701355c06.png)

rzumer replied:
Trailing new line means at the end of the file. There is none there.

Buenia0 replied:
I see. Thanks for the headsup!

Buenia0 replied:
Alright, just fixed the trailing new line issue!

rzumer commented on src/frontend/qt_sdl/EmuSettingsDialog.cpp:
```diff
@@ -36,6 +38,8 @@ extern bool RunningSomething;
 
 bool EmuSettingsDialog::needsReset = false;
 
+bool RumblePakEnabled = false;
```
This doesn't seem to have been resolved with the rest of the comments.

RSDuck commented:
what still bugs me, is that your changes aren't really integrated @rzumer's previous slot-2 work. It would probably be the best if the solar sensor and the gba cart were integrated into your scheme of slot2 additions (both from the code, but also from the UI).

That opens of course more questions, the solar sensor for example is automatically inserted for the associated games. Should that be performed with other peripherals as well? Yesterday we also had some discussions about refactoring Slot1 and Slot2 games as polymorphic classes.

Please tell us what you think! If you want to do it, should further refactorings go right into this PR or be made with separate one?

rzumer commented:
> That opens of course more questions, the solar sensor for example is automatically inserted for the associated games. Should that be performed with other peripherals as well?

I think you're conflating the GBA cartridges that enable the sensor controls when inserted, and the DS card that samples the sensor. There's no sensor "inserted" with the DS card alone.

RSDuck commented:
> I think you're conflating the GBA cartridges that enable the sensor controls when inserted, and the DS card that samples the sensor. There's no sensor "inserted" with the DS card alone. Detecting GBA cartridges that can vibrate and enabling rumbling when they are inserted would be another example of that. In most cases it doesn't make much sense to automatically insert a peripheral based on the DS card.

oh sorry, you're right. Also good point regarding GBA games which can rumble, so ideally Slot-2 peripherals and GBA games should be separated.

rzumer commented:
> oh sorry, you're right. Also good point regarding GBA games which can rumble, so ideally Slot-2 peripherals and GBA games should be separated (from a code standpoint, so that the functionality can be reused).

I think so too, but a GBA cartridge is still a slot-2 device, so it's not intuitive for it to be separated from the others in the interface.

WaluigiWare64 commented:
> I think I know why the Ubuntu aarch64 build is failing. It's because of an installation issue with the grub-efi-amd64-signed package. I have no idea what's causing that error, so if anyone has any ideas as to why that's the case, please let me know!
@Buenia0 fixed this with PR #764.

Buenia0 commented:
Thanks so much, WaluigiWare64!

WaluigiWare64 commented on src/NDS.cpp:
```diff
@@ -1916,7 +1947,7 @@ u16 ARM9Read16(u32 addr)
     switch (addr & 0xFF000000)
     {
-    case 0x02000000:
+    case 0x02000000:        
```
There seems to be extra whitespace at the end of this line.

Buenia0 replied:
Thanks for reminding me, WaluigiWare64! I'll fix it as soon as I can!

rzumer commented:
Just as a reminder there are still unaddressed issues:
* Slot-2 cartridge type is still hardcoded integers only defined in a comment, should be enum values
* GBA cartridges are still separate from all other slot-2 cartridges, which leads to unintuitive/confusing behaviour

poudink commented:
how do you want slot-2 stuff to be "merged"? like a drop down menu the emu settings menu where you can choose between using a gba rom or a slot 2 accessory?

rzumer commented:
My idea for a first implementation would be something like that, but I'm open to alternatives.

* GBA cartridge as an option in the slot-2 dropdown
* GBA cartridge either settable from the settings menu, or cartridge type set to GBA automatically when a new GBA ROM is successfully loaded from the Open... menu or drag-and-drop, with an OSD message to reflect that

This reminds me slot-2 cartridge type should be stored in save states if it is not there already.

RSDuck commented:
> how do you want slot-2 stuff to be "merged"? like a drop down menu the emu settings menu where you can choose between using a gba rom or a slot 2 accessory?

yes, correct me if I'm wrong, but I think currently you can select both a gba rom as well as a slot 2 peripheral at the same time and it's only a matter of how the things are ordered in the code what will happen.

Buenia0 commented:
> Just as a reminder there are still unaddressed issues:
> 
> * Slot-2 cartridge type is still hardcoded integers only defined in a comment, should be enum values
> * GBA cartridges are still separate from all other slot-2 cartridges, which leads to unintuitive/confusing behaviour

Currently working on resolving those right now, rzumer! Should have something ready soon!

kyandora commented:
@rzumer Are all values in GBACart(_SRAM)::DoSavestate actually used? I've been reworking GBACart trying to fit it in with the rest, but I can't fully wrap my head around the save states.

rzumer commented:
> @rzumer Are all values in GBACart(_SRAM)::DoSavestate actually used? I've been reworking GBACart trying to fit it in with the rest, but I can't fully wrap my head around the save states.

What is confusing you about it?

Save states store emulator state to restore them correctly when loaded. For the GBA cartridge the behaviour is a bit special because we can't reasonably store the full GBA cartridge ROM in the save state, both because it can be excessively large and because it can make sharing save states difficult from a legal standpoint. We also can't reasonably store a file path to a GBA ROM, because they can move around at any time and it would be annoying for a save state to not work as expected due to an associated GBA ROM having moved somewhere else. Save states are associated with DS games, so there is no such issue with DS ROMs moving around (they need to be loaded in advance anyway).

The solution I came up with was to only store the GBA cartridge header and any save data. This allows DS titles that detect inserted cartridges to identify the inserted game, and for titles that read save data (like Pokémon games) the functionality remains intact, though if anything is written to the GBA save data after loading a save state, it won't be stored back onto the disk.

Now that I think about it there are some titles that read from the GBA cartridge ROM, so this is not a perfect solution. I posted an issue about it at #796.

Buenia0 commented:
> > Just as a reminder there are still unaddressed issues:
> > 
> > * Slot-2 cartridge type is still hardcoded integers only defined in a comment, should be enum values
> > * GBA cartridges are still separate from all other slot-2 cartridges, which leads to unintuitive/confusing behaviour
> 
> Currently working on resolving those right now, rzumer! Should have something ready soon!

Quick update: a (WIP) pull request by a new contributor should fix those issues:

https://github.com/Buenia0/melonDS/pull/2

Buenia0 commented:
Quick update on this pull request: I'm currently finishing up my fourth semester of college, and am currently swamped with final projects and exams. As such, I won't be able to work on this pull request for a few weeks. But I assure you, once that semester of college is behind me, work on this pull request will pick up the pace! Have a belated Happy Thanksgiving!

Buenia0 commented:
Hello, everyone. Sorry that I haven't really worked on this PR in a while. It's just that the holidays and a bunch of various other projects of mine got in the way. But don't worry about a thing, because I'm currently working on that PR right now. Should have something on that front soon!

Buenia0 commented:
Also, the author of gbe-plus, Shonumi, actually mentioned this PR on the r/emulation subreddit.

Here's the link to the corresponding subreddit post:

[https://www.reddit.com/r/emulation/comments/iv82x1/edge_of_emulation_magic_reader/](url)

SA-X0 commented:
The recent build of this branch, no games or firmware loads.
Screen stays white at starting either.

Buenia0 commented:
Alright, SA-X0, I'm looking into it! Should have some intel on this issue for you soon...

Buenia0 commented:
Just downloaded the build in question, and oddly enough, everything's working fine on my end.

No idea why the most recent build would fail on your end...

SA-X0 commented:
Odd, the master builds no issue and past build works fine for me, well one before the recent one of this branch.
Also don't think it matters, mine is AMD.

Buenia0 commented:
You're quite welcome, my friend!

Buenia0 commented:
Hello, everyone. I just noticed that the Ubuntu aarch64 build is, for some strange reason, broken again on the latest build of this branch. Here are the relevant logs for that build:

https://github.com/Arisotura/melonDS/pull/719/checks?check_run_id=1850133456

If anyone has come up with a fix for this issue, please let me know ASAP!

Buenia0 commented:
> > > Just as a reminder there are still unaddressed issues:
> > > 
> > > * Slot-2 cartridge type is still hardcoded integers only defined in a comment, should be enum values
> > > * GBA cartridges are still separate from all other slot-2 cartridges, which leads to unintuitive/confusing behaviour
> > 
> > 
> > Currently working on resolving those right now, rzumer! Should have something ready soon!
> 
> Quick update: a (WIP) pull request by a new contributor should fix those issues:
> 
> [Buenia0#2](https://github.com/Buenia0/melonDS/pull/2)

The aforementioned pull request from earlier is finally in the review stage. Sorry for the long wait on that front!

BueniaDev commented:
Just merged my refactor of this pull request, which takes the recent cart interface redesign into account. However, I ran into a few issues with the merging process (which, by the way, were resolved), which is why GitHub Actions is claiming that 3 of the workflows need to be "approved by a maintainer with write access". It would be most appreciated if this minor issue was resolved ASAP...

WaluigiWare64 commented on src/frontend/qt_sdl/EmuSettingsDialog.cpp:
```diff
@@ -138,6 +145,7 @@ void EmuSettingsDialog::done(int r)
 
         int consoleType = ui->cbxConsoleType->currentIndex();
         int directBoot = ui->chkDirectBoot->isChecked() ? 1:0;
+	int chosenAddon = ui->cbxSlot2Addon->currentIndex();
```
Could you fix the indentation here?

WaluigiWare64 commented on src/frontend/qt_sdl/EmuSettingsDialog.cpp:
```diff
@@ -207,6 +216,8 @@ void EmuSettingsDialog::done(int r)
 
             Config::ConsoleType = consoleType;
             Config::DirectBoot = directBoot;
+	    Config::Slot2Addon = chosenAddon;
+	    NDS::LoadSlot2Addon(Config::Slot2Addon);
```
Indentation here as well.

WaluigiWare64 commented on src/frontend/qt_sdl/Input.cpp:
```diff
@@ -28,6 +28,7 @@ namespace Input
 
 int JoystickID;
 SDL_Joystick* Joystick = nullptr;
+SDL_Haptic *Haptic = nullptr;
```
Could you change this to keep with the current coding style?
```suggestion
SDL_Haptic* Haptic = nullptr;
```

WaluigiWare64 commented on src/frontend/qt_sdl/Input.h:
```diff
@@ -26,6 +26,7 @@ namespace Input
 
 extern int JoystickID;
 extern SDL_Joystick* Joystick;
+extern SDL_Haptic *Haptic;
```
Could you change this as well to keep with the current coding style?
```suggestion
extern SDL_Haptic* Haptic;
```

WaluigiWare64 commented on src/frontend/qt_sdl/Platform.cpp:
```diff
@@ -436,6 +440,25 @@ int LAN_RecvPacket(u8* data)
+{
+    if (Input::Haptic != nullptr)
+    {
+    	// TODO: Testing needed as to how long each "vibration"
+    	// of the Rumble Pak motor lasts
+    	// For now, we just run it for 48 ms
```
It's not much of an issue, but just to keep it neat, could you fix indentation?

BueniaDev commented:
Fixed all the requested changes, and as a bonus, fixed a compilation error in the non-static (i.e. DLL) builds. I hope that this PR can be merged real soon!

RSDuck commented on src/GBACart.h:
```diff
@@ -32,6 +32,7 @@ class CartCommon
     CartCommon();
     virtual ~CartCommon();
 
+    virtual bool isAddon();
```
IsAddon

RSDuck commented on src/GBACart.h:
```diff
@@ -136,11 +138,94 @@ class CartGameSolarSensor : public CartGame
+private:
+    u8 GuitarKeyStatus = 0;
+
+    u8 KeyMasks[4] = {
```
{ on the next line

RSDuck commented on src/GBACart.h:
```diff
@@ -136,11 +138,94 @@ class CartGameSolarSensor : public CartGame
+private:
+    u8 MemPakHeader[16] =
+    {
+	0xFF, 0xFF, 0x96, 0x00,
```
wrong amount of indentation

RSDuck reviewed (commented):
it's a lot better now, though gbacarts are still separate from addons on the gui side.

WaluigiWare64 commented on src/frontend/qt_sdl/EmuSettingsDialog.cpp:
```diff
@@ -158,6 +166,7 @@ void EmuSettingsDialog::done(int r)
         std::string dsiSDPath = ui->txtDSiSDPath->text().toStdString();
 
         if (consoleType != Config::ConsoleType
+	    || chosenAddon != Config::Slot2Addon
```
Indentation here,

SA-X0 commented:
Is this branch having problems?
Looks like something went wrong!

We track these errors automatically, but if the problem persists feel free to contact us. In the meantime, try refreshing.

BueniaDev commented:
There's nothing wrong, SA-X0. It's just that I'm still waiting for a few of the workflows to be approved, mainly for "security reasons". So hang tight, because I'm working on it!

BueniaDev commented:
There's nothing wrong, SA-X0. It's just that a few of the workflows require approval by a maintainer, mainly for "security reasons". So hang tight, because I'm working on it!

SA-X0 commented:
> 
> 
> There's nothing wrong, SA-X0. It's just that I'm still waiting for a few of the workflows to be approved, mainly for "security reasons". So hang tight, because I'm working on it!

Oh great, looking forward for the new builds Buenia.

BueniaDev commented:
Alright, I'm still waiting on rzumer and Rayyan to get back to me on reviewing this pull request. Not to pry or anything, but what is taking them so long?

BueniaDev commented:
Also, macOS 10.14 is deprecated on Azure Pipelines. You might wanna upgrade the macOS build-bot, like, ASAP...

nadiaholmquist commented:
> Also, macOS 10.14 is deprecated on Azure Pipelines. You might wanna upgrade the macOS build-bot, like, ASAP...

Yep that's true and I've been wanting to drop it since Apple discontinued support for it as supporting it complicates things, I thought Rayyan was the only one who could do this, that's what he had seemed to be implying in the melonDS chat at least, but doesn't seem to be the case.

I've bumped it to 10.15 and removed the workarounds we needed to build on 10.14.

RSDuck commented:
sorry, but I did review your last changes (everything afterwards seems to be just merging upstream changes) and left a comment you still haven't properly addressed:

> it's a lot better now, though gbacarts are still separate from addons on the gui side.

Arisotura is planning [some changes to the gui side of cartridges](http://melonds.kuribo64.net/comments.php?id=155), so ideally you go onto our [IRC/Discord](http://melonds.kuribo64.net/board/thread.php?id=3) to better coordinate things.

BueniaDev commented:
Hi, everyone. This is BueniaDev, the author of this pull request. Some psycho hijacked my account, and got me wrongly banned from the Kuribo64 Discord server. I'm posting this comment as proof of my identity.

GaussTek commented:
Hello, any updates on this? It would be cool to have Motion Pack support for Tony Hawk and the Hue game. By the way, when trying to download a build, it says expired. Hopefully this isn't abandoned. Thanks!

BueniaDev commented:
This is BueniaDev. YES, this pull request is still being worked on. However, we don't know much about the Motion Pak to properly implement it yet. Thanks for understanding, and I should have something ready soon!

BueniaDev commented:
Also, I think there might be an issue with the Ubuntu aarch64 buildbot. It's saying the log for that one could not be found.

BueniaDev commented:
Update: Currently working on an improved version of this pull request.

List of add-ons to be supported in the improved version:

- [ ] Rumble Pak
- [ ] Guitar Grip
- [ ] Memory Expansion Pak
- [ ] HCV-1000
- [ ] Ubisoft Thrustmaster Pedometer
- [ ] Magic Reader

CasualPokePlayer commented:
The Memory Expansion Pak is already supported upstream.

BueniaDev commented:
> The Memory Expansion Pak is already supported upstream.

Thanks for the update! I'm using that as a backbone for supporting the other addons!

BueniaDev commented:
Alright, currently working on adding support for the Rumble Pak in my improved pull-request.
In the GBACart.h header file, the following function is defined in the CartCommon class:

`virtual u32 Type();`

Can anyone please tell me what that function does, or what it's even supposed to do?
Thanks so much!

BueniaDev commented:
Never mind, I think I may have

> Alright, currently working on adding support for the Rumble Pak in my improved pull-request. In the GBACart.h header file, the following function is defined in the CartCommon class:
> 
> `virtual u32 Type();`
> 
> Can anyone please tell me what that function does, or what it's even supposed to do? Thanks so much!

Never mind, I think I might have figured that out...

BueniaDev commented:
Alright, ran into a bit of an issue with implementing the Rumble Pak. The SDL_Joystick API isn't detecting the haptic features on my Xbox One controller properly,
And apparently, according to [this comment](https://github.com/libsdl-org/SDL/issues/4435#issuecomment-896432954), the haptic API for SDL2 is deprecated for game controllers. Therefore, I am proposing that this project switch to the SDL_GameController API, which has far better support for a wider variety of game controllers.
But don't worry, I intend on doing the same for my projects as well!

BueniaDev commented:
Update: Currently implementing a workaround for the SDL_Haptic issue. Please disregard my earlier proposal.

BueniaDev commented:
Update 2: Just got the Rumble Pak support fully functional. Not everything has been committed just yet, but it should be soon!

BueniaDev commented:
Whoops, accidentally closed this branch! Reopening now...

BueniaDev commented:
Also, GitHub Actions is saying I need "a maintainer to approve running workflows" on my end. Can you please resolve this ASAP? Thanks so much!

BueniaDev commented:
Whoops, didn't realize that SDL_GameControllerHasRumble isn't available on SDL2 versions older than 2.0.18. Currently implementing a work-around for that issue...

BueniaDev commented:
Alright, I need approval on the workflows again...

patataofcourse commented:
Why do you need GitHub Actions though, can't you just self build?

BueniaDev commented:
> Why do you need GitHub Actions though, can't you just self build?

On Windows, yes, I can self-build. But on the other operating systems, I can't really do that.

BueniaDev commented:
Arrgh, GitHub Actions is requesting a maintainer's approval for the workflows on my end AGAIN! Seriously, is there any way that you could disable this requirement, so that I don't have to keep badgering a maintainer to approve the workflows on my end every time I push a new commit?

BueniaDev commented:
> Arrgh, GitHub Actions is requesting a maintainer's approval for the workflows on my end AGAIN! Seriously, is there any way that you could disable this requirement, so that I don't have to keep badgering a maintainer to approve the workflows on my end every time I push a new commit?

But seriously, can you please get somebody to approve the workflows on my end, ASAP? Thanks so much!

RSDuck commented:
I think this requirement doesn't exist for pull requests into your own fork.

BueniaDev commented:
> I think this requirement doesn't exist for pull requests into your own fork.

I see.

Also, I've noticed that the Windows build bot has been timing out when trying to install the necessary packages.
Could someone please look into that ASAP? Thanks so much!

patataofcourse commented:
built this to test it out, when i select the guitar grip this appears on the console
`GBACart: !! invalid addon type 3`

patataofcourse commented:
Something else: where's the controls for this guitar grip?

BueniaDev commented:
Yeah, this pull request is still a bit of a WIP. Just keep that in mind, okay?

BueniaDev commented:
Alright, currently adding support for the Araknoid paddle and Easy Piano addons. In addition, I also fixed the "invalid addon type" error that occurs when the Guitar Grip is selected. Not everything's been commited yet, but it should be soon!

JesseTG commented:
@BueniaDev How's this PR doing? I'd love to integrate it into [melonDS DS](https://github.com/JesseTG/melonds-ds) when it's ready.

BueniaDev commented:
This PR is still active. Haven't really gotten the chance to work on it lately, but I assure you, it will be ready by the end of this year.

JesseTG commented:
Sounds good, thanks for the update. :+1: No rush, just excited to see this come to fruition. Take as much time as you need.

BueniaDev commented:
Hello, everyone. Sorry that I haven't had time to work on this PR as much. I recently just got a new, more high-end computer, and it'll take a little while to get everything properly set up. Just wanted you to know that, OK?

MamaMelaTodo commented:
Any update?

BueniaDev commented:
I am currently working on a revamped version of this PR that incorporates better support for Slot-2 addons, as well as support for other addons. Not all the code's on GitHub yet, but it should be real soon!

BueniaDev commented:
This PR has been superseded by #1889. Closing this initial PR right now...
